src/Lib.P2P/SwarmService.cs

```diff
@@ -0,0 +1,1132 @@
+                    stream.Dispose();
+                }
+                catch (Exception)
+                {
+                    // ignore
+                }
+            }
+            finally
+            {
+                _pendingRemoteConnections.TryRemove(remote, out _);
+            }
+        }
+
+        /// <summary>
+        ///   Add a protocol that is supported by the swarm.
+        /// </summary>
+        /// <param name="protocol">
+        ///   The protocol to add.
+        /// </param>
+        public void AddProtocol(IPeerProtocol protocol)
+        {
+            lock (_protocols)
+            {
+                _protocols.Add(protocol);
+            }
+        }
+
+        /// <summary>
+        ///   Remove a protocol from the swarm.
+        /// </summary>
+        /// <param name="protocol">
+        ///   The protocol to remove.
+        /// </param>
+        public void RemoveProtocol(IPeerProtocol protocol)
+        {
+            lock (_protocols)
+            {
+                _protocols.Remove(protocol);
+            }
+        }
+
+        private void MountProtocols(PeerConnection connection)
+        {
+            lock (_protocols)
+            {
+                connection.AddProtocols(_protocols);
+            }
+        }
+
+        /// <summary>
+        ///   Stop listening on the specified <see cref="MultiAddress"/>.
+        /// </summary>
+        /// <param name="address"></param>
+        /// <returns>
+        ///   A task that represents the asynchronous operation.
+        /// </returns>
+        /// <remarks>
+        ///   Allows other peers to <see cref="ConnectAsync(MultiAddress, CancellationToken)">connect</see>
+        ///   to the <paramref name="address"/>.
+        ///   <para>
+        ///   The addresses of the <see cref="LocalPeer"/> are updated.
+        ///   </para>
+        /// </remarks>
+        public async Task StopListeningAsync(MultiAddress address)
+        {
+            if (!_listeners.TryRemove(address, out var listener))
+            {
+                return;
+            }
+
+            try
+            {
+                if (!listener.IsCancellationRequested)
+                {
+                    listener.Cancel(false);
+                }
+
+                // Remove any local peer address that depend on the cancellation token.
+                var others = _listeners
+                   .Where(l => l.Value == listener)
+                   .Select(l => l.Key)
+                   .ToArray();
+
+                LocalPeer.Addresses = LocalPeer.Addresses
+                   .Where(a => a != address)
+                   .Where(a => !others.Contains(a))
+                   .ToArray();
+
+                foreach (var other in others)
+                {
+                    _listeners.TryRemove(other, out _);
+                }
+
+                // Give some time away, so that cancel can run
+                // TODO: Would be nice to make this deterministic.
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)